physicalplan/unary/unary.go

```diff
@@ -0,0 +1,97 @@
+package unary
+
+import (
+	"context"
+	"sync"
+
+	"github.com/prometheus/prometheus/model/labels"
+	"gonum.org/v1/gonum/floats"
+
+	"github.com/thanos-community/promql-engine/physicalplan/model"
+	"github.com/thanos-community/promql-engine/worker"
+)
+
+type unaryNegation struct {
+	next model.VectorOperator
+	once sync.Once
+
+	series []labels.Labels
+
+	stepsBatch int
+	workers    worker.Group
+}
+
+func NewUnaryNegation(
+	next model.VectorOperator,
+	stepsBatch int,
+) (model.VectorOperator, error) {
+	u := &unaryNegation{
+		next:       next,
+		stepsBatch: stepsBatch,
+	}
+
+	u.workers = worker.NewGroup(stepsBatch, u.workerTask)
+	return u, nil
+}
+
+func (u *unaryNegation) Series(ctx context.Context) ([]labels.Labels, error) {
+	var err error
+	u.once.Do(func() { err = u.loadSeries(ctx) })
+	if err != nil {
+		return nil, err
+	}
+	return u.series, nil
+}
+
+func (u *unaryNegation) loadSeries(ctx context.Context) error {
+	vectorSeries, err := u.next.Series(ctx)
+	if err != nil {
+		return err
+	}
+	u.series = make([]labels.Labels, len(vectorSeries))
+	for i := range vectorSeries {
+		lbls := labels.NewBuilder(vectorSeries[i]).Del(labels.MetricName).Labels()
+		u.series[i] = lbls
+	}
+
+	u.workers.Start(ctx)
+	return nil
+}
+
+func (u *unaryNegation) GetPool() *model.VectorPool {
+	return u.next.GetPool()
+}
+
+func (u *unaryNegation) Next(ctx context.Context) ([]model.StepVector, error) {
+	in, err := u.next.Next(ctx)
+	if err != nil {
+		return nil, err
+	}
+	if in == nil {
+		return nil, nil
+	}
+	defer u.next.GetPool().PutVectors(in)
+
+	for i, vector := range in {
+		if err := u.workers[i].Send(vector); err != nil {
+			return nil, err
+		}
+	}
+
+	result := u.next.GetPool().GetVectorBatch()
+	for i, vector := range in {
+		output, err := u.workers[i].GetOutput()
+		if err != nil {
+			return nil, err
+		}
+		result = append(result, output)
+		u.next.GetPool().PutStepVector(vector)
```

Comment:
I wonder if this is safe to do here, because the input vector and output vector the same. From what I understand, `floats.Scale` modify the input vector in place. So the next operator might end up with a reference to a vector that is recycled.

I think we can improve the pooling mechanism by adding a `Recycle` method `model.StepVector`, instead of having each operator get the pool of the previous operator. This way we can propagate input vectors forward because each vector will know in which pool it needs to be put back when it is recycled.

Reply:
Should I remove this line for now?
I am not sure about how I should handle the pool part as I don't even add a vector pool for this operator. Is it required to have a pool for each op? Right now I just reuse the next pool

Reply:
I think in this case, we can reuse both the vector batch and all step vectors inside the batch. So we shouldn't need to request or recycle anything, just modify in place and pass forward. 

Returning the downstream pool in `GetPool()` is the correct thing to do, and it will ensure that the vector is returned to the correct pool from which it was allocated.

Reply:
You can also run the benchmark you added to see whether there is a significantly higher memory usage compared to the old engine.

Reply:
If it is in-place modification, then seems I don't need to allocate any new buffer and I can simply reuse the `[]StepVector` I get at the beginning? Probably not a good practice. WDYT?

```
	in, err := u.next.Next(ctx)
	if err != nil {
		return nil, err
	}
	if in == nil {
		return nil, nil
	}
	for i, vector := range in {
		if err := u.workers[i].Send(vector); err != nil {
			return nil, err
		}
	}

	for i := range in {
		// Make sure worker finishes the job.
		// Since it is in-place so no need another buffer.
		if _, err := u.workers[i].GetOutput(); err != nil {
			return nil, err
		}
	}

	return in, nil
```

Reply:
Yes, this seems like the best way to handle it.